File: aspnet-in-practice/13/VB/ProfileAPI/ProfileAPI/Code/CustomerProviders/Utilities.vb
Imports System
Imports System.Collections.Generic
Imports System.Linq
Imports System.Web
Imports System.Configuration.Provider
Imports System.Configuration
Imports System.Data
Imports System.Globalization

Namespace CustomerProviders

	Friend NotInheritable Class Utilities
		Private Sub New()
		End Sub
		Friend Shared Function FormatDateTimeForDbType(ByVal value As DateTime) As Object
			Return If((value.Year <= 1754), DirectCast(DBNull.Value, Object), DirectCast(value.ToString(DateTimeFormatInfo.InvariantInfo), Object))
		End Function

		Friend Shared Function ConvertFromCLRTypeToSqlDbType(ByVal propType As Type) As SqlDbType
			If propType.Equals(GetType(String)) Then
				Return SqlDbType.VarChar
			ElseIf propType.Equals(GetType(Int16)) OrElse propType.Equals(GetType(Int32)) Then
				Return SqlDbType.Int
			ElseIf propType.Equals(GetType(DateTime)) OrElse propType.Equals(GetType(System.Nullable(Of DateTime))) Then
				Return SqlDbType.DateTime
			ElseIf propType.Equals(GetType([Decimal])) Then
				Return SqlDbType.[Decimal]
			ElseIf propType.Equals(GetType(Double)) OrElse propType.Equals(GetType(Single)) Then
				Return SqlDbType.Float
			ElseIf propType.Equals(GetType(Boolean)) Then
				Return SqlDbType.Bit
			End If

			' add others if needed
			Return SqlDbType.[Variant]
		End Function

		Friend Shared Sub HasValidName(ByVal name As String)
			Dim legalChars As String = "_@#$"
			For i As Integer = 0 To name.Length - 1
				If Not [Char].IsLetterOrDigit(name(i)) AndAlso legalChars.IndexOf(name(i)) = -1 Then
					Throw New ProviderException("Table and column names cannot contain: " & name(i))
				End If
			Next
		End Sub
	End Class

End Namespace

File: aspnet-in-practice/13/CS/ProfileAPI/ProfileAPI/Code/CustomerProviders/Utilities.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Web;
using System.Configuration.Provider;
using System.Configuration;
using System.Data;
using System.Globalization;

namespace ProfileAPI.CustomerProviders
{
	internal static class Utilities
	{
		internal static object FormatDateTimeForDbType(DateTime value)
		{
			return (value.Year <= 1754) ? (object)DBNull.Value : (object)value.ToString(DateTimeFormatInfo.InvariantInfo);
		}

		internal static SqlDbType ConvertFromCLRTypeToSqlDbType(Type propType)
		{
			if (propType.Equals(typeof(string)))
				return SqlDbType.VarChar;
			else if (propType.Equals(typeof(Int16)) || propType.Equals(typeof(Int32)))
				return SqlDbType.Int;
			else if (propType.Equals(typeof(DateTime)) || propType.Equals(typeof(DateTime?)))
				return SqlDbType.DateTime;
			else if (propType.Equals(typeof(Decimal)))
				return SqlDbType.Decimal;
			else if (propType.Equals(typeof(double)) || propType.Equals(typeof(float)))
				return SqlDbType.Float;
			else if (propType.Equals(typeof(bool)))
				return SqlDbType.Bit;

			// add others if needed
			return SqlDbType.Variant;
		}

		internal static void HasValidName(string name)
		{
			string legalChars = "_@#$";
			for (int i = 0; i < name.Length; ++i)
			{
				if (!Char.IsLetterOrDigit(name[i]) && legalChars.IndexOf(name[i]) == -1)
					throw new ProviderException("Table and column names cannot contain: " + name[i]);
			}
		}
	}

}